### PR TITLE
[C-API] Increase ML_TENSOR_RANK_LIMIT @open sesame 04/06 11:25

### PR DESCRIPTION
**Pull request by niley7464**
This patch increases ML_TENSOR_RANK_LIMIT from 4 to 16.

Related to : https://github.com/nnstreamer/nnstreamer/pull/3987

Signed-off-by: Yelin Jeong <yelini.jeong@samsung.com>

### CODE REVIEW

**Comment by niley7464** [comment]
I assumed this condition, but I'm not familiar at version control so feel free to tell the points to fix
- This PR will be merged at the same time as nnstreamer version up ( I used 2.3.1 version for nnstreamer temporarily )
- api can check nnstreamer's version

**Review by myungjoo — APPROVED** [review]

**Review by songgot — APPROVED** [review]
Looks good to me

File: c/src/ml-api-common.c [hunk 1]
```diff
@@ -762,6 +763,25 @@ _ml_tensors_data_clone_no_alloc (const ml_tensors_data_s * data_src,
   return ML_ERROR_NONE;
 }
 
+/**
+ * @brief Gets the max tensor rank for the ml-api version
+ * @note nnstreamer supports max rank 8 since 2.3.0
```
Comment (niley7464) [inline c/src/ml-api-common.c:768]: I'm not sure which version to make rank lower.
Is it correct  rank 8 is support since 2.3.0..?

Comment (anyj0527) [reply c/src/ml-api-common.c:768]: @myungjoo @jaeyun-jung
NNStreamer would be released as v2.4.0 in Tizen 7.5. Should this line be `2.4.0`?

Comment (jaeyun-jung) [reply c/src/ml-api-common.c:768]: IMO 2.3.0 (nnstreamer v2.2.0 for Tizen 7.0 release, v2.3.x is unstable, prod version before tizen 7.5 release but now published from ppa..)

**Comment by anyj0527** [comment]
This requires https://github.com/nnstreamer/nnstreamer/pull/3987 to be merged and applied into Tizen-unified.

**Comment by again4you** [comment]
## Question
In my opinion, the backward compatibility error might occur in case the old version of the nnstreamer application is installed in the new version of Tizen.  For example, below example code should be fixed and uploaded to Tizen store again. right? If then, this code is not acceptable by our Tizen PM.
* Previous application code
```cpp
/* Below code should be fixed and re-complied */
ml_tensor_dimension dim = { CH, CAM_WIDTH, CAM_HEIGHT, 1 };
...
ml_tensors_info_create(&info);
ml_tensors_info_set_count(info, 1);
ml_tensors_info_set_tensor_type(info, 0, ML_TENSOR_TYPE_UINT8);
ml_tensors_info_set_tensor_dimension(info, 0, dim);
```

## Possible solutions
* Default: ML_TENSOR_RANK_LIMIT is 4 for backward compatibility 
* Add new API to increase the ML_TENSOR_RANK_LIMIT to 8.
  * e.g. ml_extend_rank_limit()

**Comment by niley7464** [comment]
It seems this PR has a backward compatibility error, so I closed Tizen ACR.
For not modifying example code and Tizen application, here is the possible solutions we disucssed last week.

```C
// rank <= 4, using previous api
ml_tensors_info_set_tensor_dimension(info, 0, dim);

// rank > 4, using ml_option
ml_option_h option;
ml_option_create (&option);
ml_option_set (option, "input_dimension", "a:b:c:d:e:f:g:h", NULL);
ml_tensors_info_set_nth_info (info, 0, option);
ml_single_open_with_option (&single, option);
```

**Comment by jaeyun-jung** [comment]
> ```c
> ml_option_h option;
> ml_option_create (&option);
> ml_option_set (option, "input_dimension", "a:b:c:d:e:f:g:h", NULL);
> ml_tensors_info_set_nth_info (info, 0, option);
> ```

Suggestion:
ml_option_h option;
ml_option_create (&option);

ml_option_set (option, "types", "uint8,uint8,uint8", NULL);
ml_option_set (option, "dimensions", "a:b:c:d:e:f:g:h,a:b:c:d:e:f:g:h,a:b:c:d:e:f:g:h", NULL);
ml_tensors_info_set_with_option (info, option);

ml_option_destroy (option);

File: c/src/ml-api-common.c (hunk 1 above)
Comment (myungjoo) [reply c/src/ml-api-common.c:768]: If you want a visible distinguishment, you can say "since 2.3.1".
Then, update nnstreamer version .

**Comment by again4you** [comment]
### Tizen TCT Failure list
* tct-nnstreamer-native-utc (Native)
  * utc_ml_single_get_property_p()

* Tizen.MachineLearning.Inference (.Net)
  * Tizen.MachineLearning.Inference.Tests.SingleShotTests.GetValue_NO_EXCEPTION()
  * Tizen.MachineLearning.Inference.Tests.SingleShotTests.SetValue_NO_EXCEPTION()

* tct-mlsingleshot-tizen-tests (Web)
  * SingleShot_getValue()
  * SingleShot_setValue()

* Detailed Error message
```
execute case: Tizen.MachineLearning.Inference.Tests # Tizen.MachineLearning.Inference.Tests.SingleShotTests.SetValue_NO_EXCEPTION...(FAIL)
    Failed to set input dimension\n  Expected string length 7 but was 15. Strings differ at index 7.\n  Expected: "5:1:1:1"\n  But was:  "5:1:1:1:1:1:1:1"\n  ------------------^\n
  Expected: True
  But was:  False

```

**Comment by niley7464** [comment]
@again4you, Thank you for checking!!

> ### Tizen TCT Failure list
> * tct-nnstreamer-native-utc (Native)
>   
>   * utc_ml_single_get_property_p()

Now this TC passed without change dimension string value after #283

**Comment by niley7464** [comment]
## Tizen ACR
* https://code.sec.samsung.net/jira/browse/ACR-1755

File: tests/capi/unittest_capi_datatype_consistency.cc [hunk 2]
```diff
@@ -25,9 +25,9 @@
  */
 TEST (nnstreamer_datatypes, test_all_1)
 {
-  EXPECT_EQ ((int) NNS_TENSOR_RANK_LIMIT, (int) ML_TENSOR_RANK_LIMIT);
+  EXPECT_NE ((int) NNS_TENSOR_RANK_LIMIT, (int) ML_TENSOR_RANK_LIMIT);
```
[the remark this answers is not in the document]
Comment (again4you) [reply tests/capi/unittest_capi_datatype_consistency.cc:28]: In my opinion, what about removing `EXPECT_NE()` test cases?
`EXPECT_NE()` could make users confused.

**Comment by again4you** [comment]
## Tizen TCT Failure list
* None. Good Job!

File: tests/capi/unittest_capi_datatype_consistency.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (gichan-jang) [reply tests/capi/unittest_capi_datatype_consistency.cc:28]: I agree with the @again4you. It is weird testing NE.
How about deleting the test or comparing it to the actual value?
```suggestion
  EXPECT_EQ (8, (int) NNS_TENSOR_RANK_LIMIT);
  EXPECT_EQ (16, (int) ML_TENSOR_RANK_LIMIT);
```

**Review by jaeyun-jung — APPROVED** [review]

**Comment by again4you** [comment]
I manually verified this PR.
To make a group submit with NNStreamer, this PR will be merged even though the CI test failed.

**Pull request merged**